sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -61,6 +64,38 @@ trait CheckAnalysis extends PredicateHelper {
     }
   }
 
+  private def checkColumnNames(tableDesc: CatalogTable): Unit = {
+    val colNames = tableDesc.schema.map(_.name)
+    val colNamesSet = colNames.toSet
+    checkDuplicatedColumnNames(colNames, colNamesSet, "table definition of " + tableDesc.identifier)
+
+    def requireSubsetOfSchema(subColNames: Seq[String], colType: String): Unit = {
+      val subColNamesSet = subColNames.toSet
+      checkDuplicatedColumnNames(subColNames, subColNamesSet, colType)
+      if (!subColNamesSet.subsetOf(colNamesSet)) {
+        failAnalysis(s"$colType columns (${subColNames.mkString(", ")}) must be a subset of " +
+          s"schema (${colNames.mkString(", ")}) in table '${tableDesc.identifier}'")
+      }
+    }
+
+    // Verify that the provided columns are part of the schema
+    requireSubsetOfSchema(tableDesc.partitionColumnNames, "partition")
+    requireSubsetOfSchema(tableDesc.bucketSpec.map(_.bucketColumnNames).getOrElse(Nil), "bucket")
+    requireSubsetOfSchema(tableDesc.bucketSpec.map(_.sortColumnNames).getOrElse(Nil), "sort")
+  }
+
+  private def checkDuplicatedColumnNames(
+      colNames: Seq[String],
+      colNamesSet: Set[String],
+      colType: String): Unit = {
+    if (colNamesSet.size != colNames.length) {
+      val duplicateColumns = colNames.groupBy(identity).collect {
+        case (x, ys) if ys.length > 1 => quoteIdentifier(x)
+      }
```

Comment:
we should, but the previous code doesn't consider case sensitivity either, we can do it in follow-ups.